sawtooth-sdk-transaction-processor/src/main/java/sawtooth/sdk/processor/TransactionProcessor.java

```diff
@@ -14,252 +14,278 @@
-                this.currentMessage.getCorrelationId(),
-                pingResponse.toByteString());
-            this.currentMessage = null;
-          } else if (this.currentMessage.getMessageType()
-              == Message.MessageType.TP_PROCESS_REQUEST) {
-            TransactionHandler handler = this.findHandler(this.currentMessage);
-            if (handler == null) {
-              break;
-            }
-            TransactionProcessor.process(this.currentMessage, this.stream, handler);
-            this.currentMessage = null;
+        currentMessage = this.stream.receive();
+        if (currentMessage != null) {
+          if (currentMessage.getMessageType() == Message.MessageType.PING_REQUEST) {
+            respondToPing(currentMessage);
+          } else if (currentMessage.getMessageType() == Message.MessageType.TP_PROCESS_REQUEST) {
+            submitTaskForMessage(currentMessage);
           } else {
-            LOGGER.info("Unknown Message Type: " + this.currentMessage.getMessageType());
-            this.currentMessage = null;
+            LOGGER.info("Unknown Message Type: " + currentMessage.getMessageType());
           }
         } else {
           // Disconnect
           LOGGER.info("The Validator disconnected, trying to register.");
-          this.registered = false;
-          for (int i = 0; i < this.handlers.size(); i++) {
-            TransactionHandler handler = this.handlers.get(i);
-            TpRegisterRequest registerRequest =
-                TpRegisterRequest.newBuilder()
-                    .setFamily(handler.transactionFamilyName())
-                    .addAllNamespaces(handler.getNameSpaces())
-                    .setVersion(handler.getVersion())
-                    .build();
+          this.registered.compareAndSet(true, false);
+        }
+      }
+    }
+    try {
+      flushMessages();
+      executorService.shutdown();
+      executorService.awaitTermination(DEAULT_SHUTDOWN_TIMEOUT, TimeUnit.SECONDS);
+    } catch (TimeoutException exc) {
+      // We are exiting so log the exception and go away
+      LOGGER.log(Level.FINE, exc.getMessage());
+    } catch (InterruptedException exc) {
+      LOGGER.log(Level.FINER, "Interrupted during shutdown", exc);
+    }
+  }
 
-            try {
-              Future fut =
-                  this.stream.send(
-                      Message.MessageType.TP_REGISTER_REQUEST, registerRequest.toByteString());
-              fut.getResult();
-              this.registered = true;
-            } catch (InterruptedException ie) {
-              LOGGER.log(Level.WARNING, ie.toString());
-            } catch (ValidatorConnectionError vce) {
-              LOGGER.log(Level.WARNING, vce.toString());
-            }
-          }
+  /**
+   * Send a registration request for all handlers in this processor.
+   */
+  private void registerHandlers() {
+    if (!this.registered.get()) {
+      LOGGER.info("Registering handlers");
```

Comment:
I'd probably move this to within the loop so that the handler information can be added to the logging line, mostly to avoid adding trace logging lines that lack additional context.